docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,219 @@
+
+If no eviction thresholds are specified, the `kubelet` will not do any eviction.
+
+An eviction expression is of the form:
+
+`<eviction-signal><operator><quantity>`
+
+* valid `eviction-signal` tokens are `memory.available`
+* valid `operator` tokens are `<`
+* valid `quantity` tokens must match the quantity representation used by kubernetes
+
+To configure hard eviction thresholds, the following flag will be supported:
+
+```
+--eviction-hard="": A set of eviction thresholds (e.g. memory.available<1Gi) that if met would trigger a pod eviction.
+```
+
+To configure soft eviction thresholds, the following flags will be supported:
+
+```
+--eviction-soft="": A set of eviction thresholds (e.g. memory.available<1.5Gi) that if met over a corresponding grace period would trigger a pod eviction.
+--eviction-soft-grace-period="": A set of eviction grace periods (e.g. memory.available=1m30s) that correspond to how long a soft eviction threshold must hold before triggering a pod eviction.
+```
+
+For each soft eviction threshold, a corresponding grace period must be provided.  If no grace period is provided, the kubelet
+will error on start-up.
+
+## Eviction Monitoring Interval
+
+The `kubelet` will support monitoring available compute resource usage at the same
+specified housekeeping interval as `cAdvisor` housekeeping (10s in v1.2 by default).
+
+## Eviction scenario
+
+Let's assume the operator started the `kubelet` with the following:
+
+```
+--eviction-hard="memory.available<100Mi"
+--eviction-soft="memory.available<300Mi"
+--eviction-soft-grace-period="memory.available=30s"
+```
+
+The `kubelet` will run a sync loop that looks at the available memory
+on the node as reported from `cAdvisor` by calculating (capacity - workingSet).
+If available memory is observed to drop below 100Mi, the `kubelet` will immediately
+initiate eviction. If available memory is observed as falling below `300Mi`,
+it will record when that signal was observed internally in a cache.  If at the next
+sync, that criteria was no longer satisfied, the cache is cleared for that
+signal.  If that signal is observed as being satisfied for longer than the
+specified period, the `kubelet` will initiate eviction to attempt to
+reclaim the resource that has met its eviction threshold.
+
+## Eviction of Pods
+
+If an eviction threshold has been met, the `kubelet` will initiate the
+process of evicting pods until it has observed the signal has gone below
+its defined threshold.
+
+The eviction sequence works as follows:
+
+* for each monitoring interval, if eviction thresholds have been met
+ * find candidate pod
+ * fail the pod
+ * block until pod is terminated on node
+
+The `kubelet` will respect graceful termination periods for pods when
+doing eviction.  If a pod is not terminated within its graceful termination
+period because a container does not happen to die (i.e. processes stuck in
+disk IO for example), the `kubelet` may select an additional pod to fail.
```

Comment:
Considering that the pod would be terminated abruptly by the kernel in the event of a system OOM, should we support graceful terminations here?

The amount of memory that can be sacrificed for the sake of stability will be pretty low in reality. If we were to wait for >30s to evict a pod, system OOM will most likely kick in, making this whole eviction feature mute.